File: app/src/main/java/ru/sccraft/arenacontrol/MainActivity.java
package ru.sccraft.arenacontrol;

import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;

public class MainActivity extends AppCompatActivity {

    Server[] сервер;
    String[] file;
    Fe fe;
    ListView lv;
    private boolean разрешить_использование_интендификатора = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        fe = new Fe(this);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        lv = findViewById(R.id.listView_main);

        FloatingActionButton fab = findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                //Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG).setAction("Action", null).show();
                Intent intent = new Intent(MainActivity.this, AddServerActivity.class);
                startActivity(intent);
            }
        });
        {
            String рекламаID = fe.getFile("adid");
            if (рекламаID.contains("1")) {
                разрешить_использование_интендификатора = true;
            } else {
                запросить_интендификатор();
            }
        }
    }

    private void запросить_интендификатор() {
        AlertDialog.Builder диалог = new AlertDialog.Builder(this);
        диалог.setTitle(R.string.intendificatorReqest)
                .setMessage(R.string.intendificatorMessage)
                .setCancelable(false)
                .setPositiveButton(R.string.yes, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        fe.saveFile("adid", "1");
                        разрешить_использование_интендификатора = true;
                    }
                })
                .setNegativeButton(R.string.about, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        Intent intent = new Intent(MainActivity.this, AboutActivity.class);
                        startActivity(intent);
                        finish();
                    }
                });
        диалог.show();
    }

    @Override
    protected void onResume() {
        super.onResume();
        file = fileList();
        if (file.length == 0) {
            //нет серверов
            setTitle(R.string.noServers);
            String[] ошибка = new String[]{getString(R.string.noServers)};
            ArrayAdapter<String> адаптер = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, ошибка);
            lv.setAdapter(адаптер);
            lv.setOnItemClickListener(new AdapterView.OnItemClickListener() {
                @Override
                public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                    Intent intent = new Intent(MainActivity.this, AddServerActivity.class);
                    startActivity(intent);
                }
            });
            lv.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
                @Override
                public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id) {
                    return false;
                }
            });
        } else {
            setTitle(R.string.app_name);
            {
                ArrayList<Server> pre = new ArrayList<>();
                for (String файл : file) {
                    if (!(файл.contains(".json"))) continue;
                    pre.add(Server.Companion.fromJSON(fe.getFile(файл)));
                }
                сервер = pre.toArray(new Server[pre.size()]);
            }
            SharedPreferences настройки = PreferenceManager.getDefaultSharedPreferences(this);
            String сортировать_по = настройки.getString("settings_sort_servers_by", "");
            String[] настройки_сортировки = getResources().getStringArray(R.array.settings_sort_servers_by);
            if (сортировать_по.equals(настройки_сортировки[0])) {
                Arrays.sort(сервер, new Comparator<Server>() {
                    @Override
                    public int compare(Server server, Server t1) {
                        return server.получить_токен().compareTo(t1.получить_токен());
                    }
                });
            } else if (сортировать_по.equals(настройки_сортировки[1])) {
                Arrays.sort(сервер, new Comparator<Server>() {
                    @Override
                    public int compare(Server server, Server t1) {
                        return ("" + server.получить_id()).compareTo(("" + t1.получить_id()));
                    }
                });
            } else if (сортировать_по.equals(настройки_сортировки[2])) {
                Arrays.sort(сервер, new Comparator<Server>() {
                    @Override
                    public int compare(Server server, Server t1) {
                        return server.получить_имя().compareTo(t1.получить_имя());
                    }
                });
            }
            ServerAdapter адаптер = new ServerAdapter(this, сервер);
            lv.setAdapter(адаптер);
            lv.setOnItemClickListener(new AdapterView.OnItemClickListener() {
                @Override
                public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                    if (разрешить_использование_интендификатора) {
                        Server s = сервер[position];
                        Intent intent = new Intent(MainActivity.this, UpdateActivity.class);
                        intent.putExtra("server", s.toJSON());
                        startActivityForResult(intent, 1);
                    } else {
                        запросить_интендификатор();
                    }
                }
            });
            lv.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
                @Override
                public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id) {
                    String s = сервер[position].toJSON();
                    Intent intent = new Intent(MainActivity.this, ServerActivity.class);
                    intent.putExtra("server", s);
                    startActivity(intent);
                    return true;
                }
            });
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        switch (id) {
            case R.id.action_about:
                Intent intent = new Intent(MainActivity.this, AboutActivity.class);
                startActivity(intent);
                return true;
            case R.id.action_settings:
                Intent intent1 = new Intent(MainActivity.this, SettingsActivity.class);
                startActivity(intent1);
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == 1) {
            if (resultCode == 0) {
                String сервер = data.getStringExtra("server");
                Intent intent = new Intent(MainActivity.this, ServerActivity.class);
                intent.putExtra("server", сервер);
                startActivity(intent);
            }
        }
    }
}

File: app/src/main/java/ru/sccraft/arenacontrol/MainActivity.kt
package ru.sccraft.arenacontrol

import android.content.Intent
import android.os.Bundle
import android.preference.PreferenceManager
import android.view.Menu
import android.view.MenuItem
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.ListView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import com.google.android.material.floatingactionbutton.FloatingActionButton
import java.util.*

class MainActivity : AppCompatActivity() {

    internal lateinit var сервер: Array<Server>
    internal lateinit var file: Array<String>
    internal lateinit var fe: Fe
    internal lateinit var lv: ListView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        fe = Fe(this)
        setContentView(R.layout.activity_main)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        lv = findViewById(R.id.listView_main)

        val fab = findViewById<FloatingActionButton>(R.id.fab)
        fab.setOnClickListener {
            //Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG).setAction("Action", null).show();
            val intent = Intent(this@MainActivity, AddServerActivity::class.java)
            startActivity(intent)
        }
        run {
        }
    }

    override fun onResume() {
        super.onResume()
        file = fileList()
        if (file.size == 0 || (file.size == 1 && file[0].contains("PersistedInstallation"))) {
            //нет серверов
            setTitle(R.string.noServers)
            val ошибка = arrayOf(getString(R.string.noServers))
            val адаптер = ArrayAdapter(this, android.R.layout.simple_list_item_1, ошибка)
            lv.adapter = адаптер
            lv.onItemClickListener = AdapterView.OnItemClickListener { parent, view, position, id ->
                val intent = Intent(this@MainActivity, AddServerActivity::class.java)
                startActivity(intent)
            }
            lv.onItemLongClickListener = AdapterView.OnItemLongClickListener { parent, view, position, id -> false }
        } else {
            setTitle(R.string.app_name)
            run {
                val pre = ArrayList<Server>()
                for (файл in file) {
                    if (!файл.contains(".json")) continue
                    if (файл.contains("PersistedInstallation")) continue
                    pre.add(Server.fromJSON(fe.getFile(файл)))
                }
                сервер = pre.toTypedArray()
            }
            val настройки = PreferenceManager.getDefaultSharedPreferences(this)
            val сортировать_по = настройки.getString("settings_sort_servers_by", "")
            val настройки_сортировки = resources.getStringArray(R.array.settings_sort_servers_by)
            if (сортировать_по == настройки_сортировки[0]) {
                Arrays.sort(сервер) { server, t1 -> server.получить_токен().compareTo(t1.получить_токен()) }
            } else if (сортировать_по == настройки_сортировки[1]) {
                Arrays.sort(сервер) { server, t1 -> ("" + server.получить_id()).compareTo("" + t1.получить_id()) }
            } else if (сортировать_по == настройки_сортировки[2]) {
                Arrays.sort(сервер) { server, t1 -> server.получить_имя().compareTo(t1.получить_имя()) }
            }
            val адаптер = ServerAdapter(this, сервер)
            lv.adapter = адаптер
            lv.onItemClickListener = AdapterView.OnItemClickListener { parent, view, position, id ->
                val s = сервер[position]
                val intent = Intent(this@MainActivity, UpdateActivity::class.java)
                intent.putExtra("server", s.toJSON())
                startActivityForResult(intent, 1)
            }
            lv.onItemLongClickListener = AdapterView.OnItemLongClickListener { parent, view, position, id ->
                val s = сервер[position].toJSON()
                val intent = Intent(this@MainActivity, ServerActivity::class.java)
                intent.putExtra("server", s)
                startActivity(intent)
                true
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId


        when (id) {
            R.id.action_about -> {
                val intent = Intent(this@MainActivity, AboutActivity::class.java)
                startActivity(intent)
                return true
            }
            R.id.action_settings -> {
                val intent1 = Intent(this@MainActivity, SettingsActivity::class.java)
                startActivity(intent1)
            }
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == 1) {
            if (resultCode == 0) {
                val сервер = data!!.getStringExtra("server")
                val intent = Intent(this@MainActivity, ServerActivity::class.java)
                intent.putExtra("server", сервер)
                startActivity(intent)
            }
        }
    }
}
